WireDomain/Sources/WireDomain/NotificationService/NotificationService.swift

```diff
@@ -0,0 +1,189 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UserNotifications
+import WireDataModel
+import WireLogging
+
+/// Receives push notifications, process the pending events through the `NotificationSession` to generate a notification
+/// content based on these events.
+final class NotificationService: UNNotificationServiceExtension {
+
+    // MARK: - Failure
+
+    enum Failure: Error {
+        case missingSelfClientID
+        case notAuthenticated
+    }
+
+    // MARK: - Properties
+
+    private let logger = WireLogger.notifications
+    private var notificationSession: NotificationSession?
+    private var contentHandler: ((UNNotificationContent) -> Void)?
+    private var onGoingTask: Task<Void, Never>?
+
+    // MARK: - Object lifecycle
+
+    override init() {
+        logger.info("initializing notification service")
+        super.init()
+    }
+
+    // MARK: - Notifications
+
+    override func didReceive(
+        _ request: UNNotificationRequest,
+        withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void
+    ) {
+        onGoingTask?.cancel()
+        self.contentHandler = contentHandler
+
+        onGoingTask = Task {
+            do {
+                let notificationUserInfo = request.content.userInfo
+
+                let notification = try NotificationPayload(
+                    userInfo: notificationUserInfo
+                )
+
+                notificationSession = try await createNotificationSession(
+                    userID: notification.userID
+                )
+
+                try await notificationSession?.processPushNotification(
+                    eventID: notification.eventID
+                )
```

Comment:
I'm wondering if we need to retain the content handler. Do you think it could be passed to the session and the session takes care of invoking it?